File: design_patterns/singleton/singleton.java

public class Singleton {

    private static Singleton instance = null;

    public int value = 0;

    public static Singleton getInstance() {
        if (instance == null) {
            instance = new Singleton();
        }

        return instance;
    }
    
    private Singleton() {}

    public static void main(String[] args) {

        Singleton s1 = Singleton.getInstance();
        Singleton s2 = Singleton.getInstance();

        System.out.println("s1 value: " + s1.value);
        
        s1.value = 4;

        System.out.println("s2 value: " + s2.value);
    }
}

File: design_patterns/singleton/singleton.cpp
#include <iostream>
#include <memory>

class Singleton {
private:

    static std::unique_ptr<Singleton> instance;

    Singleton() = default;
    Singleton(const Singleton& src) = delete;
    Singleton& operator=(const Singleton& rhs) = delete;

public:
    virtual ~Singleton() = default;


    static Singleton &getInstance() {
        if (instance == nullptr) {
            instance.reset(new Singleton);
        }

        return *instance;
    }

    int value = 0;
};

std::unique_ptr<Singleton> Singleton::instance;

int main() {

    Singleton& s1 = Singleton::getInstance();
    Singleton& s2 = Singleton::getInstance();

    std::cout << "s1 value: " << s1.value << "\n";

    s1.value = 4;

    std::cout << "s2 value: " << s2.value << "\n";

    return 0;
}
